src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -219,6 +249,144 @@ public void finish()
+    {
+        SortedSet<String> indexNames = new TreeSet<>();
+        try
+        {
+            while (true)
+            {
+                List<SSTableIndex> referencedIndexes = new ArrayList<>();
+                boolean failed = false;
+
+                Map<Expression, NavigableSet<SSTableIndex>> view = getView(expressions);
+
+                for (SSTableIndex index : view.values().stream().flatMap(Collection::stream).collect(Collectors.toList()))
+                {
+                    indexNames.add(index.getIndexContext().getIndexName());
+
+                    if (index.reference())
+                    {
+                        referencedIndexes.add(index);
+                    }
+                    else
+                    {
+                        failed = true;
+                        break;
+                    }
+                }
+
+                if (failed)
+                {
+                    // TODO: This might be a good candidate for a table/index group metric in the future...
+                    referencedIndexes.forEach(QueryController::releaseQuietly);
+                }
+                else
+                {
+                    return view;
+                }
+            }
+        }
+        finally
+        {
+            Tracing.trace("Querying storage-attached indexes {}", indexNames);
+        }
+    }
+
+    private Map<Expression, NavigableSet<SSTableIndex>> getView(Collection<Expression> expressions)
+    {
+        // first let's determine the primary expression
+        Pair<Expression, NavigableSet<SSTableIndex>> primary = calculatePrimary(expressions);
+
+        Map<Expression, NavigableSet<SSTableIndex>> indexes = new HashMap<>();
+        for (Expression expression : expressions)
+        {
+            // Non-index column query should only act as FILTER BY for satisfiedBy(Row) method
+            // because otherwise it likely to go through the whole index.
+            if (expression.context.isNotIndexed())
+            {
+                continue;
+            }
+
+            // primary expression, we'll have to add as is
+            if (primary != null && expression.equals(primary.left))
+            {
+                indexes.put(primary.left, primary.right);
+
+                continue;
+            }
+
+            View view = expression.context.getView();
+
+            NavigableSet<SSTableIndex> readers = new TreeSet<>(SSTableIndex.COMPARATOR);
+            if (primary != null)
+            {
+                for (SSTableIndex index : primary.right)
+                    readers.addAll(view.match(index.minKey(), index.maxKey()));
+            }
+            else
+            {
+                readers.addAll(applyScope(view.match(expression)));
+            }
+
+            indexes.put(expression, readers);
+        }
+
+        return indexes;
+    }
+
+    // The purpose of this method is to, attempt to, calculate the most selective expression based on
```

Comment:
```suggestion
    // The purpose of this method is to attempt to calculate the most selective expression based on
```
...or I guess you could like put "attempt to" in parenthesis if you wanted it to be an aside.